### PR TITLE
Re-apply MySQL net write/read settings changeset

### LINKED ISSUE #58553: MySQL compatibility: SET net_write_timeout
### Description
Required for AWS QuickSight SPICE import.

Sample query:

```sql
SET net_write_timeout = 60
```

### Expected behavior

Either ignored completely or translated to [send_timeout](https://clickhouse.com/docs/en/operations/settings/settings#connect-timeout-receive-timeout-send-timeout) (?)

See also: https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_net_write_timeout

### PR DESCRIPTION
**Pull request by slvrtrn**
Re-apply changeset from (PR) https://github.com/ClickHouse/ClickHouse/pull/58835, which was reverted in (PR) https://github.com/ClickHouse/ClickHouse/pull/58872 due to (issue) https://github.com/ClickHouse/ClickHouse/issues/58918. 

Since the issue is primarily "cosmetic" (wrong `limit` value in `system.settings`, but it's actually applied) and (PR) https://github.com/ClickHouse/ClickHouse/pull/59018 is still a draft, I rewrote an offending test, so it is not blocking and maybe we can still include it in 24.1.

---

Original description from (PR) https://github.com/ClickHouse/ClickHouse/pull/58835:

Resolves https://github.com/ClickHouse/ClickHouse/issues/58553

Allows to use the following settings via MySQL interface:

```sql
SET NET_WRITE_TIMEOUT = X;
SET NET_READ_TIMEOUT = Y;
```

`net_write_timeout` is translated into the native `send_timeout` ClickHouse setting and, similarly, `net_read_timeout` into `receive_timeout`. These statements work regardless of the input query being upper or lower case.

Additionally, fixed an issue where changing MySQL `SQL_SELECT_LIMIT` setting was possible only if the entire statement was in upper case.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
MySQL interface gained support for `net_write_timeout` and `net_read_timeout` settings. `net_write_timeout` is translated into the native `send_timeout` ClickHouse setting and, similarly, `net_read_timeout` into `receive_timeout`. Fixed an issue where it was possible to set MySQL `sql_select_limit` setting only if the entire statement was in upper case.

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 7d785c47e292b8f28d7c1b947393432316cd9c14 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59293/7d785c47e292b8f28d7c1b947393432316cd9c14/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59293/7d785c47e292b8f28d7c1b947393432316cd9c14/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Server/MySQLHandler.cpp [hunk 1]
```diff
@@ -66,7 +66,7 @@ static String showCountWarningsReplacementQuery(const String & query);
 static String selectEmptyReplacementQuery(const String & query);
 static String showTableStatusReplacementQuery(const String & query);
 static String killConnectionIdReplacementQuery(const String & query);
-static String selectLimitReplacementQuery(const String & query);
+static std::optional<String> setSettingReplacementQuery(const String & query, const String & mysql_setting, const String & native_setting);
```
Comment (rschu1ze) [inline src/Server/MySQLHandler.cpp:69]: If you move the implementation of these methods before where they are used, you would not need to forward-declare them (l. 64-69).

File: src/Server/MySQLHandler.cpp [hunk 2]
```diff
@@ -601,12 +617,12 @@ static String showTableStatusReplacementQuery(const String & query)
     return query;
 }
 
-static String selectLimitReplacementQuery(const String & query)
+static std::optional<String> setSettingReplacementQuery(const String & query, const String & mysql_setting, const String & native_setting)
```
Comment (rschu1ze) [inline src/Server/MySQLHandler.cpp:620]: Could we rename `native_setting` to `clickhouse_setting`? (just to remove ambiguity a bit)

File: src/Server/MySQLHandler.cpp [hunk 3]
```diff
@@ -601,12 +617,12 @@ static String showTableStatusReplacementQuery(const String & query)
...
 {
-    const String prefix = "SET SQL_SELECT_LIMIT";
-    if (query.starts_with(prefix))
-        return "SET limit" + std::string(query.data() + prefix.length());
-    return query;
+    const String prefix = "SET " + mysql_setting;
+    if (0 == strncasecmp(prefix.c_str(), query.c_str(), prefix.size()))
```
Comment (rschu1ze) [inline src/Server/MySQLHandler.cpp:623]: Can we please not use rather low-level C functions in C++ code (whereas `strncasecmp()` isn't even standardized)? To compare two strings case-insensitively, see e.g. [here](https://stackoverflow.com/questions/11635/case-insensitive-string-comparison-in-c).

File: tests/queries/0_stateless/02967_mysql_settings_override.sh [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+. "$CURDIR"/../shell_config.sh
+
+CHANGED_SETTINGS_QUERY="SELECT name, value FROM system.settings WHERE name IN ('send_timeout', 'receive_timeout') AND changed;"
+
+TEST_TABLE="mysql_settings_override_test"
+
+DROP_TABLE="DROP TABLE IF EXISTS $TEST_TABLE;"
+CREATE_TABLE="CREATE TABLE $TEST_TABLE (s String) ENGINE MergeTree ORDER BY s;"
+INSERT_STMT="INSERT INTO $TEST_TABLE VALUES ('a'), ('b'), ('c'), ('d');"
+SELECT_STMT="SELECT * FROM $TEST_TABLE;"
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02967_mysql_settings_override.sh:18]: Better put in an `ORDER BY s` at the end of the `SELECT` to guarantee a stable result order.

File: tests/queries/0_stateless/02967_mysql_settings_override.sh [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+echo "-- Init"
+${MYSQL_CLIENT} --execute "$DROP_TABLE $CREATE_TABLE $INSERT_STMT $SELECT_STMT" # should fetch all 4 records
+
+echo "-- Uppercase tests"
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02967_mysql_settings_override.sh:23]: Maybe `echo "-- Uppercase setting name"`?

File: tests/queries/0_stateless/02967_mysql_settings_override.sh [hunk 6]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
+# Tags: no-fasttest
+# Tag no-fasttest: requires mysql client
+
+# Tests the override of certain MySQL proprietary settings to ClickHouse native settings
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02967_mysql_settings_override.sh:5]: Tests that certain MySQL-proprietary settings are mapped to ClickHouse-native settings.

File: src/Server/MySQLHandler.h [hunk 7]
```diff
@@ -92,9 +92,12 @@ class MySQLHandler : public Poco::Net::TCPServerConnection
...
     std::unique_ptr<Session> session;
 
-    using ReplacementFn = std::function<String(const String & query)>;
-    using Replacements = std::unordered_map<std::string, ReplacementFn>;
-    Replacements replacements;
+    using QueryReplacementFn = std::function<String(const String & query)>;
+    using QueriesReplacements = std::unordered_map<std::string, QueryReplacementFn>;
+    QueriesReplacements queries_replacements;
+
+    using SettingsReplacements = std::unordered_map<std::string, std::string>;
```
Comment (rschu1ze) [inline src/Server/MySQLHandler.h:99]: It is not immediately clear what is mapped to what, could you add a comment to l. 98, e.g.

`/// MySQL setting name --> ClickHouse setting name`

File: src/Server/MySQLHandler.cpp [hunk 8]
```diff
@@ -342,16 +344,30 @@ void MySQLHandler::comQuery(ReadBuffer & payload, bool binary_protocol)
         bool should_replace = false;
         bool with_output = false;
 
-        for (auto const & x : replacements)
+        // Queries replacements
+        for (auto const & [query_to_replace, replacement_fn] : queries_replacements)
         {
-            if (0 == strncasecmp(x.first.c_str(), query.c_str(), x.first.size()))
+            if (0 == strncasecmp(query_to_replace.c_str(), query.c_str(), query_to_replace.size()))
```
Comment (rschu1ze) [inline src/Server/MySQLHandler.cpp:350]: (please see my other comment about `strncasecmp`)

**Comment by slvrtrn** [comment]
@rschu1ze, I addressed the review comments. Can you please have a look one more time?

**Pull request merged**